neqo-http3/src/send_message.rs

```diff
@@ -213,37 +150,20 @@ impl SendStream for SendMessage {
     /// `TransportStreamDoesNotExist` if the transport stream does not exist (this may happen if `process_output`
     /// has not been called when needed, and HTTP3 layer has not picked up the info that the stream has been closed.)
     fn send(&mut self, conn: &mut Connection) -> Res<()> {
-        let label = if ::log::log_enabled!(::log::Level::Debug) {
-            format!("{}", self)
-        } else {
-            String::new()
-        };
-
-        if let SendMessageState::SendingInitialMessage { ref mut buf, fin } = self.state {
-            let sent = Error::map_error(
-                conn.stream_send(self.stream_id, buf),
-                Error::HttpInternal(5),
-            )?;
-            qlog::h3_data_moved_down(&mut conn.qlog_mut(), self.stream_id, sent);
-
-            qtrace!([label], "{} bytes sent", sent);
-
-            if sent == buf.len() {
-                if fin {
-                    Error::map_error(
-                        conn.stream_close_send(self.stream_id),
-                        Error::HttpInternal(6),
-                    )?;
-                    self.state = SendMessageState::Closed;
-                    qtrace!([label], "done sending request");
-                } else {
-                    self.state = SendMessageState::SendingData;
-                    self.conn_events.data_writable(self.stream_id);
-                    qtrace!([label], "change to state SendingData");
-                }
+        let sent = Error::map_error(self.stream.send_buffer(conn), Error::HttpInternal(5))?;
+        qlog::h3_data_moved_down(&mut conn.qlog_mut(), self.stream_id(), sent);
+
+        qtrace!([self], "{} bytes sent", sent);
+
+        if !self.stream.has_buffered_data() {
+            if self.fin {
+                Error::map_error(
+                    conn.stream_close_send(self.stream_id()),
+                    Error::HttpInternal(6),
+                )?;
+                qtrace!([self], "done sending request");
             } else {
-                let b = buf.split_off(sent);
-                *buf = b;
+                self.conn_events.data_writable(self.stream_id());
```

Comment:
Why don't we call self.stream_writable() here?

There's a corner case here that I'm not sure about: if we buffer the headers and that takes *all* the remaining flow control window for the stream, we'll get to this point and the stream will not be writable.  Do we need to ask the connection if there is space available?

Reply:
This is fine. DataWritable is just a signal to try writing and if it fails it is fine. This can already happen in other ways, e.g. if session flow control opens up, all active stream will get this signal, but first streams that get the signal can eat up all the credits and later streams will not be able to write data.

Reply:
A comment to that effect would be helpful.